### PR TITLE
Enhance `RedirectingHttpRequesterFilter`: redirect headers and message body

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

With pre-existing implementation of `RedirectingHttpRequesterFilter` its
hard for users to redirect headers and a message body. There is an
opportunity to improve API for request customization and automatically
redirect those components for relative redirects.

Modifications:

- Introduce `RedirectConfig` and `RedirectConfigBuilder` for extensive
redirect configuration;
- Automatically redirect headers and message body for relative locations;
- Allow configuring what methods should be redirected;
- Let users provide a custom predicate to avoid redirects in some
conditions;
- Make "change POST -> GET" behavior opt-in;
- Give users access to make any changes for the redirecting request,
keeping the context of the original request and redirect response;
- Deprecate `MultiAddressHttpClientBuilder#maxRedirects`;
- Add `MultiAddressHttpClientBuilder#followRedirects(RedirectConfig)`;
- Add `RedirectingHttpRequesterFilter(RedirectConfig)` ctor;
- Deprecate unnecessary ctors for `RedirectingHttpRequesterFilter`;
- Update tests and improve test coverage;
- Update redirect examples;

Result:

Full automatic handling for relative redirects, extensible API to
configure non-relative redirects, simplified user experience.

### CODE REVIEW

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java [hunk 1]
```diff
@@ -227,75 +250,176 @@ private boolean shouldRedirect(final int redirectCount, final StreamingHttpRespo
...
+            // NOTE: for security reasons we do not copy any headers or payload body from the original request by
+            // default for non-relative redirects.
+            for (CharSequence headerName : redirectSingle.config.headersToRedirect) {
+                for (CharSequence headerValue : request.headers().values(headerName)) {
+                    redirectRequest.addHeader(headerName, headerValue);
+                }
+            }
+
+            final Config config = redirectSingle.config;
+            if (config.redirectPayloadBody) {
```
Comment (Scottmitch) [inline servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:364]: should this be conditional based upon the request/response?

**Review by Scottmitch — COMMENTED** [review]
looking good! few comments/questions.

File: servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/RedirectWithStateUrlClient.java [hunk 2]
```diff
@@ -15,95 +15,37 @@
...
 public final class RedirectWithStateUrlClient {
 
-    private static final AsyncContextMap.Key<StreamingHttpRequest> ORIGINAL_REQUEST = newKey();
-
     public static void main(String... args) throws Exception {
         try (HttpClient client = HttpClients.forMultiAddressUrl()
-                // This is an optional configuration that applies more restrictive limit for the number or redirects:
-                .maxRedirects(1)
+                // Enables redirection with a custom configuration:
+                .followRedirects(config -> config.maxRedirects(3)
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/RedirectWithStateUrlClient.java:44]: No, `MultiAddressHttpClientBuilder#maxRedirects(int)` was not the reason, it's deprecated and will go away. Actual reasons:
1. `Consumer` of `RedirectConfiguration` API allows me to pre-configure something (`allowNonRelativeRedirects(true)` in this case) before I give it to users. I followed an approach of `SingleAddressInitializer` instead of `SslConfig` to be consistent at `MultiAddressHttpClientBuilder` API.
2. It also gives me an opportunity to have a single place for javadoc of all these options, instead of duplicating javadoc between `RedirectConfig` and `RedirectConfigBuilder`.
3. Proposed approach requires `RedirectConfigBuilder` to be in `http-api` module. Otherwise, it's weird if `MultiAddressHttpClientBuilder#followRedirects` requires `RedirectConfig` but there is no provided API to build that config. But the filter is in `http-utils`. It complicates the filter if `RedirectConfigBuilder` is too far away from the filter and limits how the redirecting filter can extend options.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+     * it performs a non-relative redirect.
+     *
+     * @param headerNames Names of headers that have to be copied on each non-relative redirect
+     * @return {@code this}
+     * @see #allowNonRelativeRedirects(boolean)
+     * @see #redirectPayloadBody(boolean)
+     * @see #trailersToRedirect(CharSequence...)
+     * @see #prepareRequest(TriConsumer)
+     */
+    RedirectConfiguration headersToRedirect(CharSequence... headerNames);
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java:116]: If users need more flexibility and control, they can use `prepareRequest` that gives access to the original request and redirect request

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java [hunk 4]
```diff
@@ -227,75 +250,176 @@ private boolean shouldRedirect(final int redirectCount, final StreamingHttpRespo
...
-        private static StreamingHttpRequest prepareRedirectRequest(final StreamingHttpRequest request,
-                                                                   final StreamingHttpResponse response,
-                                                                   final StreamingHttpRequestFactory requestFactory) {
-            final HttpRequestMethod method = defineRedirectMethod(request.method());
+        private StreamingHttpRequest prepareRedirectRequest(final StreamingHttpRequest request,
+                                                            final StreamingHttpResponse response,
+                                                            final StreamingHttpRequestFactory requestFactory) {
+            final HttpRequestMethod method = defineRedirectMethod(response.status().code(), request.method());
             final CharSequence locationHeader = response.headers().get(LOCATION);
             assert locationHeader != null;
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:282]: We never reach this point if `location` header is null, `shouldRedirect` method ensures this

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java (hunk 1 above)
Comment (idelpivnitskiy) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:364]: If users need more flexibility and control, they can use `prepareRequest` that gives access to the original request, redirect response, and redirect request. This API is intended to be simple for cases when they know what has to be copied.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java:117]: WDYT, should we also add `redirectAllHeaders/redirectAllTrailers` to make it even easier or it may open extra risk (users won't think what headers are sensitive)? They can also do:
```java
.prepareRequest((prevRequest, redirectResponse, redirectedRequest) -> {
    redirectedRequest.setHeaders(prevRequest.headers());
}))
```

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java [hunk 5]
```diff
@@ -117,13 +150,30 @@ public RedirectingHttpRequesterFilter(final boolean onlyRelativeClient,
...
     public RedirectingHttpRequesterFilter(final boolean onlyRelativeClient,
                                           final boolean onlyRelativeConnection,
                                           final int maxRedirects) {
-        this.onlyRelativeClient = onlyRelativeClient;
-        this.onlyRelativeConnection = onlyRelativeConnection;
-        this.maxRedirects = maxRedirects;
+        this(maxRedirects, onlyRelativeClient);
+    }
+
+    private RedirectingHttpRequesterFilter(final int maxRedirects,
```
Comment (chemicL) [inline servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java:162]: Perhaps using the existing public constructor would reduce the total amount of constructors as in the end that public constructor can be transformed to a private one. Also, the other new private constructor could flip the order of arguments to reduce the ambiguity level that having first boolean paramter mean different things depending on the chosen constructor.

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java [hunk 6]
```diff
@@ -162,26 +211,129 @@ public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnect
...
 
     private Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
                                                   final HttpExecutionStrategy strategy,
                                                   final StreamingHttpRequest request,
-                                                  final boolean onlyRelative) {
-        final Single<StreamingHttpResponse> response = delegate.request(strategy, request);
-        if (maxRedirects <= 0) {
+                                                  final boolean allowNonRelativeRedirects) {
+        final Single<StreamingHttpResponse> response = delegate.request(strategy,
+                // Duplicate each payload buffer chunk to allow safely replaying it without worry that indexes can move
```
Comment (chemicL) [inline servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java:224]: This improvement seems unrelated to the context of this PR – perhaps can be introduced as a separate PR?

**Review by chemicL — CHANGES_REQUESTED** [review]

File: servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                        .maxRedirects(3)
+                        // by default, POST requests don't follow redirects:
+                        .allowedMethods(GET, POST)
+                        // apply additional restrictions which redirects to follow:
+                        .shouldRedirect((request, response) -> {
+                            String location = response.headers().get(LOCATION).toString();
+                            // allow only:
+                            return location.startsWith("/") // relative redirects
+                                    // OR non-relative redirects to a trusted server:
+                                    || location.startsWith("https://localhost:" + SECURE_SERVER_PORT);
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java:63]: The goal of this example is to demonstrate at least one use-case how `shouldRedirect` function can be used. Users can go further in its implementation, which is out of scope of this example. If you have ideas of necessary changes, I'm happy to incorporate those.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java [hunk 8]
```diff
@@ -105,9 +105,18 @@ public StreamingHttpClient buildStreaming() {
                             defaultReqRespFactory(buildContext.httpConfig().asReadOnly(),
                                     executionContext.bufferAllocator())));
 
-            // Need to wrap the top level client (group) in order for non-relative redirects to work
-            urlClient = maxRedirects <= 0 ? urlClient :
-                    new RedirectingHttpRequesterFilter(false, maxRedirects).create(urlClient);
+            if (maxRedirects > 0 || followRedirectsInitializer != null) {
+                final RedirectingHttpRequesterFilter.Builder redirectBuilder =
+                        new RedirectingHttpRequesterFilter.Builder().allowNonRelativeRedirects(true);
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:110]: We pre-set the default value for users because we know that this client is able to follow non-relative redirects. This is to keep pre-existing behavior and to be consistent with other 3-party clients (like AHC). If users want to disable non-relative redirects, they can override this default via configuration.

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java [hunk 9]
```diff
@@ -130,54 +140,73 @@ public void onSubscribe(final Cancellable cancellable) {
...
-                    }
-                    target.onSuccess(result);
-                    return;
-                } else if (newScheme != null) {
-                    // Rewrite absolute-form location to relative-form request-target in case only relative redirects
-                    // are supported and this is a relative redirect
+
+            final String newScheme = newRequest.scheme();
+            if (isRelative(request, scheme, newRequest)) {
+                if (redirectSingle.allowNonRelativeRedirects && newScheme == null && scheme != null) {
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:175]: The comment one line below describes motivation. Because this filter can be applied at a multi-address level when `allowNonRelativeRedirects == true`, we have to meet its requirements. Multi-address client requires users to provide full URL (even for relative redirects) because it always parses the request-target to identify which single-address client to use.

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MultiAddressUrlHttpClientTest.java [hunk 10]
```diff
@@ -98,6 +96,7 @@ static void beforeClass() throws Exception {
         afterClassCloseables = newCompositeCloseable();
 
         client = afterClassCloseables.append(HttpClients.forMultiAddressUrl()
+                .followRedirects(config -> { /* noop */ })
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MultiAddressUrlHttpClientTest.java:99]: To enable redirects, this PR disables them by default (feedback from users that this is safer default).

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java (hunk 6 above)
Comment (idelpivnitskiy) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java:224]: This PR introduces config options that allow users to carry the payload body forward for each redirect. Without this change, some tests failing when new options are used. So, this is a related change.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+     * @param prepareRequest {@link TriConsumer} that modifies a request for redirect. First argument is the previous
+     * request, the second argument is the received response, the third argument is an initialized request for redirect
+     * @return {@code this}
+     * @see #allowNonRelativeRedirects(boolean)
+     * @see #headersToRedirect(CharSequence...)
+     * @see #redirectPayloadBody(boolean)
+     * @see #trailersToRedirect(CharSequence...)
+     */
+    RedirectConfiguration prepareRequest(
+            TriConsumer<StreamingHttpRequest, StreamingHttpResponse, StreamingHttpRequest> prepareRequest);
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java]: LMK what you think about added `RedirectRequestTransformer` interface. Open for name suggestions

File: servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (chemicL) [reply servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java:63]: I was thinking in terms of being the user, it'd be beneficial to be able to call something like `RedirectUtils.isRelative(request, response)` (or `response.isLocalRedirect()` for that matter) where internally ST is able to figure out, e.g. that `response.headers().get(LOCATION)` returning `127.0.0.1` is in fact in the DNS cache resolved from `localhost` and would provide the user with a simple answer. I don't have a solution in place, just asking if we can simplify it for the users.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (chemicL) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:110]: I'd love that to be added as a comment to this line.

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (chemicL) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:176]: Thanks for the explanation. Makes sense.

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java (hunk 6 above)
Comment (chemicL) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectingHttpRequesterFilter.java:218]: Thanks for explaining!

File: servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (chemicL) [reply servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java:63]: As a matter of fact, the recent change to the `prepareRequest` shows it's completely feasible - the api for `shouldRedirect` could also take the `relative` parameter in the functional interface.

**Review by chemicL — APPROVED** [review]

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java]: Yes, it makes everything crystal clear now :)

File: servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/RedirectWithStateUrlClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/RedirectWithStateUrlClient.java]: Followed you suggestion, fixed in this commit: https://github.com/apple/servicetalk/pull/1792/commits/cc2a1116c1abc0827568d2bac6e72fa2e1d7a444

File: servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java:65]: Added to a new interface `ShouldRedirectPredicate`: https://github.com/apple/servicetalk/pull/1792/commits/11d687a90dcd99a98c68f0127ced028a56931932

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java]: After the recent change https://github.com/apple/servicetalk/pull/1792/commits/cc2a1116c1abc0827568d2bac6e72fa2e1d7a444 users of multi-address client have to always provide their own `RedirectConfig`. So, we don't need tp pre-set any defaults for them.

**Comment by idelpivnitskiy** [comment]
Does anyone have an idea why the JDK8 build fails with:
```
> Task :servicetalk-http-api:javadoc
/home/runner/work/servicetalk/servicetalk/servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java:205: warning - Tag @see: can't find redirectRequestTransformer(RedirectRequestTransformer) in io.servicetalk.http.api.RedirectConfigBuilder
/home/runner/work/servicetalk/servicetalk/servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java:224: warning - Tag @see: can't find redirectRequestTransformer(RedirectRequestTransformer) in io.servicetalk.http.api.RedirectConfigBuilder
/home/runner/work/servicetalk/servicetalk/servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java:119: warning - Tag @link: can't find redirectRequestTransformer(RedirectRequestTransformer) in io.servicetalk.http.api.RedirectConfigBuilder
/home/runner/work/servicetalk/servicetalk/servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java:179: warning - Tag @link: can't find redirectRequestTransformer(RedirectRequestTransformer) in io.servicetalk.http.api.RedirectConfigBuilder
/home/runner/work/servicetalk/servicetalk/servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java:205: warning - Tag @link: can't find redirectRequestTransformer(RedirectRequestTransformer) in io.servicetalk.http.api.RedirectConfigBuilder
/home/runner/work/servicetalk/servicetalk/servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java:224: warning - Tag @link: can't find redirectRequestTransformer(RedirectRequestTransformer) in io.servicetalk.http.api.RedirectConfigBuilder
8 warnings

> Task :servicetalk-http-api:javadoc FAILED
```
? Works fine on JDK11 and Intellj IDEA doesn't show any warnings.

**Comment by bondolo** [comment]
Quite mysterious. You could try to use a fully qualified reference and see if that also fails. Occasionally, I have seen Javadoc get confused with colliding symbol names in nested interfaces/classes and there is more than one `redirectRequestTransformer` symbol it appears.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfig.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+     * Maximum number of redirects to follow.
+     *
+     * @return maximum number of redirects to follow.
+     */
+    int maxRedirects();
+
+    /**
+     * {@link HttpRequestMethod}s that are allowed to follow redirects.
```
Comment (Scottmitch) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfig.java:80]: the current implementation requires the list is sorted, and doesn't verify this constraint. consider the following options:
- clarify this constraint in the javadoc, in `RedirectSingle` check if they are not sorted and throw
- in `RedirectSingle` just sort them if they are not sorted.
- make this a responsibility of `RedirectRequestTransformer` which can internalize sorting requirement

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+            redirectRequest.transformMessageBody(p -> p.ignoreElements().concat(originalRequest.messageBody()));
+            // Use `transform` to update PayloadInfo flags, assuming trailers may be included in the message body
+            redirectRequest.transform(NOOP_TRAILERS_TRANSFORMER);
+            // FIXME: instead of `transform`, preserve original PayloadInfo/FlushStrategy when it's API is available
+        }
+
+        private void safeCopy(final StreamingHttpRequest request, final StreamingHttpRequest redirectRequest) {
+            // NOTE: for security reasons we do not copy any headers or payload body from the original request by
+            // default for non-relative redirects.
+            for (CharSequence headerName : headersToRedirect) {
```
Comment (Scottmitch) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java:364]: optimization/feel free to ignore for now

if `headersToRedirect.length > request.headers().size()` it may make sense to flip the iteration order, and if that is the case you can also use a `HttpHeaders` as a case-insensitive associating array to do the lookup:

```java
            // cache me in the constructor
            final HttpHeaders hdrsToRedirect = DefaultHttpHeadersFactory.INSTANCE.newHeaders();
            for (CharSequence name : headersToRedirect) {
                hdrsToRedirect.add(name, "anything");
            }
            // cache me in the constructor
            
            for (Map.Entry<CharSequence, CharSequence> hdr : request.headers()) {
                if (hdrsToRedirect.contains(hdr.getKey())) {
                    redirectRequest.addHeader(hdr.getKey(), hdr.getValue());
                }
            }
```

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java [hunk 14]
```diff
@@ -69,16 +71,15 @@
...
 final class DefaultMultiAddressUrlHttpClientBuilder
         implements MultiAddressHttpClientBuilder<HostAndPort, InetSocketAddress> {
-    // https://tools.ietf.org/html/rfc2068#section-10.3 says:
-    // A user agent SHOULD NOT automatically redirect a request more than 5 times,
-    // since such redirects usually indicate an infinite loop.
-    private static final int DEFAULT_MAX_REDIRECTS = 5;
 
     private static final String HTTPS_SCHEME = HTTPS.toString();
+    private static final RedirectConfig DEFAULT_REDIRECT_CONFIG = new RedirectConfigBuilder()
+            .allowNonRelativeRedirects(true).build();   // FIXME: remove for 0.42 branch to make redirects opt-in
```
Comment (Scottmitch) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:77]: `FIXME` -> this PR is on the main branch ... is the intention to remove this in a followup PR and just here to make back porting easier?

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:266]: consider changing `shouldRedirect` to return the `LOCATION` header if successful (or null if not) and pass into this method to clarify this relationship (and avoid double lookup).

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java [hunk 15]
```diff
@@ -198,22 +213,16 @@ public void onError(final Throwable t) {
...
         }
 
-        private boolean shouldRedirect(final int redirectCount, final StreamingHttpResponse response,
-                                       final HttpRequestMethod originalMethod) {
-            final int statusCode = response.status().code();
+        private boolean shouldRedirect(final int redirectCount, final HttpRequestMethod requestMethod,
+                                       final HttpResponseMetaData responseMetaData) {
+            final int statusCode = responseMetaData.status().code();
 
             if (statusCode < 300 || statusCode > 308) {
```
Comment (Scottmitch) [inline servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:220]: consider simplifying the control flow in this method:

```java
            if (statusCode < 300 || statusCode > 308 ||
                    (statusCode >= 304 && statusCode <= 306)) {
                // We start without support for status codes outside of this range
                // re-visit when we need to support payloads in redirects.

                // https://tools.ietf.org/html/rfc7232#section-4.1
                // The 304 (Not Modified) status code indicates that the cache value is still fresh and can be used.

                // https://tools.ietf.org/html/rfc7231#section-6.4.5
                // The 305 (Use Proxy) status code has been deprecated due to
                // security concerns regarding in-band configuration of a proxy.

                // https://tools.ietf.org/html/rfc7231#section-6.4.6
                // The 306 (Unused) status code is no longer used, and the code is reserved.
                return false;
            }
            final RedirectConfig config = redirectSingle.config;
            if (redirectCount >= config.maxRedirects()) {
                LOGGER.debug("Maximum number of redirects ({}) reached for original request: {}",
                        config.maxRedirects(), redirectSingle.originalRequest);
                return false;
            }

            if (binarySearch(config.allowedMethods(), requestMethod) < 0) {
                LOGGER.debug("Configuration does not allow redirect of method: {}", requestMethod);
                return false;
            }

            final CharSequence locationHeader = responseMetaData.headers().get(LOCATION);
            if (locationHeader == null || locationHeader.length() == 0) {
                LOGGER.debug("No location header in redirect response: {}", responseMetaData);
                return false;
            }

            return true;
```

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java [hunk 16]
```diff
@@ -130,57 +125,77 @@ public void onSubscribe(final Cancellable cancellable) {
...
+                        newRequest.requestTarget(
+                                absoluteToRelativeFormRequestTarget(newRequest.requestTarget(), newScheme));
+                    }
+                }
+
+                newRequest = redirectSingle.config.redirectRequestTransformer()
+                        .apply(relative, request, response, newRequest);
+
+                if (LOGGER.isTraceEnabled()) {
+                    LOGGER.trace("Executing redirect to '{}' for request '{}'", response.headers().get(LOCATION),
```
Comment (Scottmitch) [inline servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:185]: nit: if shouldRedirect to return the LOCATION as suggested in https://github.com/apple/servicetalk/pull/1792/files#r707906058 you could also use it here too.

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java [hunk 17]
```diff
@@ -130,57 +125,77 @@ public void onSubscribe(final Cancellable cancellable) {
...
+
+                terminalDelivered = true;   // Mark as "delivered" because we do not own `target` from this point
+                toSource(response.messageBody().ignoreElements()    // Consume any payload of the redirect response
+                        .concat(redirectSingle.requester.request(redirectSingle.strategy, newRequest)))
+                        .subscribe(new RedirectSubscriber(target, redirectSingle, newRequest, redirectCount + 1,
+                                sequentialCancellable));
+            } catch (Throwable cause) {
+                if (!terminalDelivered) {
+                    target.onError(cause);
+                }
```
Comment (Scottmitch) [inline servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java:197]: else consider logging similar to the `SubscriberUtils.safeOnSuccess` method.

**Review by Scottmitch — APPROVED** [review]
nice improvements!

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfiguration.java]: discussed offline ... I like the `DefaultRedirectRequestTransformer` approach to consolidate responsibility and give more flexibility🏅

File: servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java (hunk 1 above)
Comment (Scottmitch) [reply servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java]: this is now in `RedirectRequestTransformer` so users have control 💪 🏅

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfig.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Predicate to make the final decision if redirect should be performed or not based on the given context.
+     * <p>
+     * Implementations should prefer running this predicate as the last check, after validation of the
+     * {@link HttpResponseStatus}, {@link RedirectConfig#maxRedirects()}, {@link RedirectConfig#allowedMethods()},
+     * presence of the {@link HttpHeaderNames#LOCATION Location} header, and
+     * {@link RedirectConfig#allowNonRelativeRedirects()}.
+     */
+    @FunctionalInterface
+    interface ShouldRedirectPredicate {
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfig.java]: Renamed this one and changed to use `#redirectRequestTransformer(RedirectConfig.RedirectRequestTransformer)` in javadoc.

File: servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+ * payload body of the original request while redirecting to non-relative locations.
+ * <p>
+ * For security reasons, request methods other than {@link HttpRequestMethod#GET GET} or
+ * {@link HttpRequestMethod#HEAD HEAD}, headers and message body are not automatically redirected for non-relative
+ * locations. Users have to explicitly configure what should be redirected when they are sure that redirect does not
+ * forward to a malicious target server. Relative redirects always carry forward headers and message body. For more
+ * information, see {@link MultiAddressHttpClientBuilder#followRedirects(RedirectConfig)} and
+ * {@link RedirectConfigBuilder}.
+ */
+public final class MultiAddressRedirectClient {
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-examples/http/redirects/src/main/java/io/servicetalk/examples/http/redirects/MultiAddressRedirectClient.java]: This is consistent with factory names on `HttpClients` to help understand which example relates to which factory. I've renamed `MultiAddressRedirectClient` to `MultiAddressUrlRedirectClient`.

Also, changed the order of client example classes in docs to start with `SingleAddressRedirectClient`, and added more clarity in javadoc for each example.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfig.java (hunk 12 above)
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfig.java:80]: Good point, I've changed it to `Set` to avoid any ambiguity and clarify on the API that duplicates are not expected. Not sure if a `SortedSet` will provide any value.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java (hunk 14 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:77]: Yes, easier backporting and intention to preserve sequential git history
